soccer/gameplay/tactics/defensive_forward.py

```diff
@@ -0,0 +1,216 @@
+import main
+import robocup
+import behavior
+import constants
+import enum
+import math
+
+import composite_behavior
+import evaluation.ball
+import evaluation.defensive_positioning
+import skills.mark
+import skills.capture
+
+
+# Finds the positions to place each defender
+class DefensiveForward(composite_behavior.CompositeBehavior):
+    class State(enum.Enum):
+        # Everyone is defending
+        blocking = 1
+        # Closest defender is collecting, all others are blocking
+        collecting = 2
+
+    def __init__(self):
+        super().__init__(continuous=True)
+
+        # Robot queue
+        self.defenders = [None, None, None]
+        # Single robot collecting
+        self.collector = None
+
+        self.zone_def_pos = robocup.Point(0, 0)
+        self.mark_bots = [None, None]
+
+        self.block_dist = 0.01
+        self.block_angle_coeff = 0.5
+        self.names = ['mark_main', 'mark_sub', 'mark_zone']
+        # Distance offset for path estimation
+        self.dodge_dist = .2
+
+        self.time_since_call = 0
+        # Estimates defensive positions every X ticks
+        self.cache_amnt = 5
+
+        for s in DefensiveForward.State:
+            self.add_state(s, behavior.Behavior.State.running)
+
+        self.add_transition(behavior.Behavior.State.start,
+                            DefensiveForward.State.blocking, lambda: True,
+                            'immediately')
+
+        self.add_transition(DefensiveForward.State.blocking,
+                            DefensiveForward.State.collecting,
+                            lambda: self.within_range(), 'Collecting')
+
+        self.add_transition(
+            DefensiveForward.State.collecting, DefensiveForward.State.blocking,
+            lambda: not self.within_range(), 'Back to blocking')
+
+        self.add_transition(
+            DefensiveForward.State.collecting,
+            behavior.Behavior.State.completed,
+            lambda: self.collector is not None and evaluation.ball.robot_has_ball(self.collector.robot),
```

Comment:
I'm fairly sure it is, so this statement can be fixed by adding a check for `*.robot is not None`